### DISCUSSION TITLE
Custom regex for docker images containing semver and date and digest

### DISCUSSION THREAD

**Discussion by ocofaigh** [renovatebot/renovate#17879] [Request Help] [answered]
I am trying to get renovate to bump the following docker image tag inside a helm values file:
```
image: 
  repository: "ext/logdna-agent" 
  tag: "2.2.5-20220124.144652989e60"
  registry: icr.io
```

As there is extra text after the sem ver number, I am guessing I need some kind of custom regex package rule to allow renovate to bump it, however I am unsure the syntax to use.

Here is a snippet of the naming convention used for the images:
```
icr.io/ext/logdna-agent   2.2.5-20211227.d3e30c8c2bfc                                  9f48b49a8c44   ext         8 months ago    120 MB   -   
icr.io/ext/logdna-agent   2.2.5-20220110.093ccf3907c4                                  3fc21ac3daf7   ext         8 months ago    120 MB   -   
icr.io/ext/logdna-agent   2.2.5-20220117.4f5131557c8a                                  f6274dd34d68   ext         8 months ago    120 MB   -   
icr.io/ext/logdna-agent   2.2.5-20220124.144652989e60                                  6601478b70de   ext         7 months ago    120 MB   -   
icr.io/ext/logdna-agent   2.2.6-20220211.171e7a05c22e                                  47e890e7315a   ext         7 months ago    120 MB   -   
icr.io/ext/logdna-agent   2.2.6-20220214.af9271d6599f                                  2707c3315980   ext         7 months ago    120 MB   -   
icr.io/ext/logdna-agent   2.2.6-20220221.87fddfdd0738                                  42dddd29392a   ext         7 months ago    120 MB   -   
icr.io/ext/logdna-agent   2.2.6-20220228.dc6063a1e1b7                                  bd64275065e0   ext         6 months ago    120 MB   -   
icr.io/ext/logdna-agent   2.2.6-20220307.f3f2dd4f230d                                  080d642d510e   ext         6 months ago    120 MB   - 
```

Has anyone solved the regex for a similar use case?

**Comment by Gabriel-Ladzaretti** [comment]
I don't think we have a versioning module where these tags would be considered valid.
So I not sure a regex manager would be much of a help here.

even if there would be such versioning, differentiating between these might be tricky 
```
icr.io/ext/logdna-agent   2.2.6-20220214.af9271d6599f                                  2707c3315980   ext         7 months ago    120 MB   -   
icr.io/ext/logdna-agent   2.2.6-20220221.87fddfdd0738                                  42dddd29392a   ext         7 months ago    120 MB   -   
icr.io/ext/logdna-agent   2.2.6-20220228.dc6063a1e1b7                                  bd64275065e0   ext         6 months ago    120 MB   -   
icr.io/ext/logdna-agent   2.2.6-20220307.f3f2dd4f230d                                  080d642d510e   ext         6 months ago    120 MB   - 
```

That said, you can still probably bump the `semVer` version but I guess that's not what you are after.

**Reply by mikaello** [reply]
Unless there is something I don't get, I think we have a quite similar case, and with minor changes this is how we have solved it:

```json
  "packageRules": [
    {
      "matchPackageNames": ["icr.io/ext/logdna-agent"],
      "versioning": "regex:^(?<major>\\d+)[.](?<minor>\\d+)[.](?<patch>\\d+)-(?<build>.+)$"
    }
  ]
```

This will parse this version `2.2.6-20220307.f3f2dd4f230d` as follows:

* Major: `2`
* Minor: `2`
* Patch: `6`
* Build: `20220307.f3f2dd4f230d`

If you get multiple builds per day, e.g. `2.2.6-20220307.f3f2dd4f230d` and `2.2.6-20220307.a5d2aa6be33b`, I am not sure how they are handled, if they are alphabetically sorted by build or not. Our dependency is not released that often, so it has not been a problem so far.

**Reply by Gabriel-Ladzaretti** [reply]
I wasn't aware of Regex versioning,  thanks!

**Reply by viceice** [reply]
you should remove the hash from build, as it will be parsed as number, so it can probably cause odd issues

**Reply by ocofaigh** [reply]
@mikaello Thank you so much, this works perfectly for me!
```
  "packageRules": [
    {
      "matchPackageNames": ["icr.io/ext/logdna-agent"],
      "versioning": "regex:^(?<major>\\d+)[.](?<minor>\\d+)[.](?<patch>\\d+)-(?<build>.+)$"
    }
  ]
```

**Reply by ocofaigh** [reply]
@mikaello I noticed that with this custom versioning regex set, renovate bumped to version `3.6.0-beta.2`. Do you know off the top of your head what the best way is to tell renovate not to bump to beta builds?

**Reply by mikaello** [reply]
You could modify the last part of the regex, `(?<build>.+)` to somehow not match what you recently got, e.g. by hardcoding to the format you showed us previously, and if you also remove the hash as suggested by viceice, you end up with this:

```js
"regex:^(?<major>\\d+)[.](?<minor>\\d+)[.](?<patch>\\d+)-(?<build>\\d{8})\\.\\w+$"
```

**Comment by ocofaigh** [comment, marked as answer]
Thank you to @mikaello the solution is in https://github.com/renovatebot/renovate/discussions/17879#discussioncomment-3696276
